packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -55,10 +60,291 @@ enum _ScaffoldSlot {
+  /// 
+  /// The animation should both start and end with a value of 1.0.
+  Animation<double> getScaleAnimation({@required Animation<double> parent});
+
+  /// Animates the rotation of [Scaffold.floatingActionButton].
+  /// 
+  /// The animation should both start and end with a value of 0.0 or 1.0.
+  /// 
+  /// The animation values are a fraction of a full circle, with 0.0 and 1.0
+  /// corresponding to 0 and 360 degrees, while 0.5 corresponds to 180 degrees.
+  Animation<double> getRotationAnimation({@required Animation<double> parent});
+
+  /// Gets the value to restart a motion animation at if it's interrupted.
+  /// 
+  /// An interruption is triggered if the [Scaffold] is given a new [FloatingActionButtonPositioner]
+  /// while it is still animating a transition between two previous [FloatingActionButtonPositioner]s.
+  /// 
+  /// A sensible default is usually 0.0, which is the same as restarting
+  /// the animation from the beginning.
+  double getAnimationRestart(double previousValue) => 0.0;
+
+  @override
+  String toString() => '$runtimeType';
+}
+
+class _ScalingFabMotionAnimator extends FloatingActionButtonAnimator {
+  const _ScalingFabMotionAnimator();
+
+  @override
+  Offset getOffset({Offset begin, Offset end, double progress}) {
+    if (progress < 0.5) {
+      return begin;
+    } else {
+      return end;
+    }
+  }
+
+  @override
+  Animation<double> getScaleAnimation({Animation<double> parent}) {
+    // Animate the scale down from 1 to 0 in the first half of the animation
+    // then from 0 back to 1 in the second half.
+    final Curve curve = const Interval(0.5, 1.0, curve: Curves.ease);
+    return new AnimationMax<double>(
+      new ReverseAnimation(new CurveTween(curve: curve.flipped).animate(parent)),
+      new CurveTween(curve: curve).animate(parent),
+    );
+  }
+
+  @override
+  Animation<double> getRotationAnimation({Animation<double> parent}) {
+    // Because we only see the last half of the rotation tween, 
+    // it needs to go twice as far.
+    final Tween<double> rotationTween = new Tween<double>(
+      begin: 1.0 - _kFloatingActionButtonTurnInterval * 2, 
+      end: 1.0,
+    );
+    // This rotation will turn on the way in, but not on the way out.
+    return new AnimationMax<double>(
+      rotationTween.animate(parent),
+      new ReverseAnimation(new CurveTween(curve: const Threshold(0.5)).animate(parent)),
+    );
+  }
+
+  // If the animation was just starting, we'll continue from where we left off.
+  // If the animation was finishing, we'll treat it as if we were starting at that point in reverse.
+  // This avoids a size jump during the animation.
+  @override
+  double getAnimationRestart(double previousValue) => math.min(1.0 - previousValue, previousValue);
+}
+
+/// The geometry of the [Scaffold] before it finishes laying out.
+/// 
+/// The Scaffold passes this prelayout geometry to its [FloatingActionButtonPositioner].
```

Comment:
You need to explain what "before it finishes laying out" actually means. I think what you means is that everything in [ScaffoldGeometry] has a size and a position by the FAB (of course) does not?